### PR TITLE
Events Functionality

### LINKED ISSUE #2: [FR2] Show List of Available Events
### User Story Statement

**As a** user,
**I want** to be able view a list of available events, 
**so that** I can select a specific event and book it.

### Acceptance Criteria

- [ ] Given this is a ticket reservation application
  When I reserve a ticket for an event
  Then I can view a list of available events.

### LINKED ISSUE #4: [FR4] Reservation Cancellation
### User Story Statement

**As a** user,
**I want** to be able to cancel reservations I have already made, 
**so that** I can express that I am unable to attend.

### Acceptance Criteria

- [ ] Given this is a ticket reservation application
  When I want to express that I would not attend an event
  Then I can cancel a reservation I have already made

### LINKED ISSUE #6: [FR6] Add New Event
### Admin Story Statement

**As an** admin,
**I want** to be able to add a new event, 
**so that** I can show to the user new events that are currently available.

### Acceptance Criteria

- [ ] Given this is a ticket reservation application
  When I want to show to the user new events that are currently available
  Then I can add a new event

### LINKED ISSUE #7: [FR7] Edit an Existing Event
### Admin Story Statement

**As an** admin,
**I want** to be able to edit an existing event, 
**so that** I can update existing events with modified information.

### Acceptance Criteria

- [ ] Given this is a ticket reservation application
  When I want to update existing events with modified information
  Then I can edit an existing event

### LINKED ISSUE #8: [FR8] Cancel an Event
### Admin Story Statement

**As an** admin,
**I want** to cancel existing events, 
**so that** I can notify the users that an event is no longer available.

### Acceptance Criteria

- [ ] Given this is a ticket reservation application
  When I want to notify users that an event is no longer available
  Then I can cancel an existing event

### LINKED ISSUE #10: [FR2-T] Show List of Available Events
### Related User Story

#2 

### Preconditions

- The application is running
- The user has access to the app

### Acceptance Criteria

- Given this is a ticket reservation application
When I reserve a ticket for an event
Then I can view a list of available events.

### PR DESCRIPTION
**Pull request by silvehh**
## Added
Admins:
- Can add events
- can view events
- can modify events
- can cancel events

Users:
- can view events

Added tests for all that functionality

## Modified
- A test was failing because test was calling firestore API before firestore was loaded, so added a waitforview with a 10 second timeout
- some tests failed to teardown properly and caused subsequent tests to fail when running full test suite at once\
- modified user tests to track and delete created users at the end of tests to avoid database accumulation

### CODE REVIEW

File: app/src/main/java/com/example/ticketreservationapp/AddEventActivity.java [hunk 1]
```diff
@@ -39,17 +48,64 @@ protected void onCreate(Bundle savedInstanceState) {
...
 
         btnPublish.setOnClickListener(v -> {
             if (formHelper.validateFields()) {
                 Event event = formHelper.createEvent();
-                Toast.makeText(this, "Event published!", Toast.LENGTH_SHORT).show();
-                if (event != null) {
-                    finish();
+                if (eventToEdit != null) {
+                    event.setId(eventToEdit.getId());
+                    rw.updateEvent(event).addOnSuccessListener(aVoid -> {
```
Comment (A-Eskaf) [inline app/src/main/java/com/example/ticketreservationapp/AddEventActivity.java:70]: Edit flow updates when eventToEdit != null, but doesn’t verify eventToEdit.getId() != null (cancel flow already does this).
Suggested fix: guard update with eventToEdit.getId() != null and show a user-facing error if ID is missing.

File: app/src/main/java/com/example/ticketreservationapp/EventListActivity.java [hunk 2]
```diff
@@ -3,44 +3,155 @@
...
+            filteredEvents.addAll(allEvents);
+        } else {
+            for (Event event : allEvents) {
+                if (event.getEventName().toLowerCase().contains(text.toLowerCase())) {
+                    filteredEvents.add(event);
+                }
+            }
+        }
+        adapter.notifyDataSetChanged();
+    }
```
Comment (A-Eskaf) [inline app/src/main/java/com/example/ticketreservationapp/EventListActivity.java:138]: event.getEventName().toLowerCase() can throw NullPointerException if Firestore returns an event with a null/missing eventName.
Suggested fix: null-check before calling toLowerCase() (or default to empty string).

**Review by A-Eskaf — CHANGES_REQUESTED** [review]
Overall this is a strong PR with meaningful functionality and solid test additions. The event flow is much more complete now (add/edit/cancel/list/search), and the structure is clear.

Before merge, I’d request 2 small fixes:
1. Add a null-safe check in event search (`EventListActivity`) to avoid a possible crash when `eventName` is missing.
2. Add an ID guard in edit-update flow (`AddEventActivity`) so update doesn’t run with a null event ID.

Once those are patched, this looks good to merge.

**Review by A-Eskaf — APPROVED** [review]

**Comment by A-Eskaf** [comment]
Great work on the last two commits:

1. **Firestore Configuration Refactor** (`0e95575`) - Nice cleanup by centralizing the Firestore emulator setup in `TestUtils.getTestFirestore()`. This eliminates code duplication and ensures consistent test configuration across all test files.

2. **Event Validation Enhancement** (`add0e30`) - Solid improvement to input validation in `EventFormHelper` with proper numeric checks for ticket count and price fields. The new `isWholeNumber()` and `isValidDecimal()` helper methods are clean, and the corresponding tests cover the edge cases well.

The error handling in `AddEventActivity` with the try-catch for `NumberFormatException` adds a good safety net.
